### PR TITLE
deep copy from pod.Annotations to avoid concurrent map read and write

### LINKED ISSUE #110084: kubelet: fix to prevent concurrent map read and map
### What happened?

When we add custom annotations to PodSandboxConfig.Annotations in function generatePodSandboxConfig， kubelet crashes with errror:
kubelet[2257292]: fatal error: concurrent map read and map write

The role of function generatePodSandboxConfig(located at kuberuntime_sandbox.go) is to create SandboxConfig that is used to RunSandbox. In some scenarios, we want to pass custom annotations to runtime  through PodsandboxConfig.Annotations .

In kubelet,  multiple goroutins access *v1.Pod fetched from podManager at the same time with no lock. For example, syncPod goroutin and  volumeManager goroutin get *v1.Pod from podManager， when there is  writing pod.Annotations operations,  kubelet will panic.

So, in function generatePodSandboxConfig，a deep copy of pod.Annotations  should be performed.

### What did you expect to happen?

Kubelet process won't panic when we add custom annotations to PodSandboxConfig.

### How can we reproduce it (as minimally and precisely as possible)?

Writing podSandboxConfig.Annotations in function generatePodSandboxConfig 

### Anything else we need to know?

_No response_

### Kubernetes version

kubelet  v1.20.0

### Cloud provider

None

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by gaohuatao-1**
In kubelet, multiple goroutins access *v1.Pod fetched from podManager at the same time with no lock. For example, syncPod goroutin and volumeManager goroutin get *v1.Pod from podManager， when there is writing pod.Annotations operations, kubelet will panic.

/kind bug

Fixes [#110085 ](https://github.com/kubernetes/kubernetes/issues/110084)

### CODE REVIEW

File: pkg/kubelet/kuberuntime/labels.go [hunk 1]
```diff
@@ -88,8 +88,14 @@ func newPodLabels(pod *v1.Pod) map[string]string {
 }
 
 // newPodAnnotations creates pod annotations from v1.Pod.
+// Deep copy from pod.Annotations to avoid concurrent map read and map write
 func newPodAnnotations(pod *v1.Pod) map[string]string {
-	return pod.Annotations
+	anno := make(map[string]string)
```
Comment (FillZpp) [inline pkg/kubelet/kuberuntime/labels.go:93]: `make(map[string]string, len(pod.Annotations))`

Comment (gaohuatao-1) [reply pkg/kubelet/kuberuntime/labels.go]: OK

File: pkg/kubelet/kuberuntime/labels.go [hunk 2]
```diff
@@ -88,8 +88,14 @@ func newPodLabels(pod *v1.Pod) map[string]string {
 }
 
 // newPodAnnotations creates pod annotations from v1.Pod.
+// Deep copy from pod.Annotations to avoid concurrent map read and map write
```
Comment (Abirdcfly) [inline pkg/kubelet/kuberuntime/labels.go:91]: Nit: Like other similar logic, it seems that the comments here can be removed.

Comment (gaohuatao-1) [reply pkg/kubelet/kuberuntime/labels.go]: Good idea, the comments have been removed.

**Comment by gaohuatao-1** [comment]
Who can add approve label?

**Comment by endocrimes** [comment]
/close

as per the discussion in the original issue, the panic only occurs in a fork, and we don't expect mutations in this part of the kubelet today. The right place to handle this right now is in the fork.

That can be revisited in the future, but for now I'd like to avoid changing behaviour.

**Pull request closed without merging**